### PR TITLE
ENG-11434 voltadmin shutdown/pause timeout

### PR DESCRIPTION
**Pull request by zeemanhe**
added a mechanism to monitor the draining of outstanding transactions.  The command will be timed out if no transactions have been drained within timeout period after the last successfully drained transaction.

### CODE REVIEW

File: lib/python/voltcli/checkstats.py [hunk 1]
```diff
@@ -269,7 +340,49 @@ def check_command_log(runner):
...
+        lastUpdatedTime = monitorStatisticsProgress(lastValidationParamms, currentValidationParams, lastUpdatedTime, runner, 'COMMANDLOG')
+        lastValidationParamms = [outstandingByte, outstandingTxn]
         time.sleep(1)
+
+def monitorStatisticsProgress(lastUpdatedParams, currentParams, lastUpdatedTime, runner, component):
+    currentTime = time.time()
+
+    timeout = -1;
+    if runner.opts.timeout:
+        timeout = (runner.opts.timeout) * 60
```
Comment (xinjiacs) [inline lib/python/voltcli/checkstats.py]: Do we have any negative timeout input checking?

File: lib/python/voltcli/checkstats.py [hunk 2]
```diff
@@ -108,6 +142,10 @@ def dr_producer_stats(runner, partition_min_host, partition_min, partition_max):
             last_acked = -1
         else:
             last_acked = r[9]
+
+        if last_queued == -1 and last_acked == -1:
```
Comment (xinjiacs) [inline lib/python/voltcli/checkstats.py]: What's the case for -1? Can you add comments for it?

Comment (zeemanhe) [inline lib/python/voltcli/checkstats.py]: no transactions have been registered  for the partition in the queue yet.

File: lib/python/voltcli/checkstats.py [hunk 3]
```diff
@@ -351,15 +345,7 @@ def check_command_log(runner):
...
-
-    timeout = -1;
-    if runner.opts.timeout:
-        timeout = (runner.opts.timeout) * 60
-
-    #no timeout check
-    if timeout == -1:
-        return currentTime
-
+    timeout = (runner.opts.timeout) * 60
```
Comment (xinjiacs) [inline lib/python/voltcli/checkstats.py]: What if the user does not specify the timeout option? Previously you handled this situation.

File: lib/python/voltcli/voltadmin.d/pause.py [hunk 4]
```diff
@@ -25,10 +25,13 @@
         VOLT.BooleanOption('-w', '--wait', 'waiting',
                            'wait for all DR and Export transactions to be externally processed',
                            default = False),
-        VOLT.IntegerOption('-t', '--timeout', 'timeout', 'The timeout value in minute if @Statistics is not progressing.', default = 2),
+        VOLT.IntegerOption('-t', '--timeout', 'timeout', 'The timeout value in minutes if @Statistics is not progressing.', default = 2),
     )
 )
 def pause(runner):
+    if runner.opts.timeout and runner.opts.timeout < 0:
+        runner.abort_with_help('You cannot specify timeout less than zero minutes.')
```
Comment (xinjiacs) [inline lib/python/voltcli/voltadmin.d/pause.py]: Maybe better with "Timeout value less than zero is invalid."

Does it make sense to have a 0 value timeout? Do we want to give the unit in SECONDS?

Comment (zeemanhe) [inline lib/python/voltcli/voltadmin.d/pause.py]: updated with seconds, default=120.

File: lib/python/voltcli/checkstats.py [hunk 5]
```diff
@@ -70,23 +76,45 @@ def check_dr_producer(runner):
...
+    partitionMinProgressed = cmp(lastPartitionMin, currentPartitionMin)
+    partitionMaxprogressed = cmp(lastPartitionMax, currentPartitionMax)
+    #stats moved
+    if partitionMinProgressed <> 0 or partitionMaxprogressed <> 0:
+        return currentTime
+
+    timeSinceLastUpdate = currentTime - lastUpdatedTime
+    #stats timeout
+    if timeSinceLastUpdate > timeout:
+         msg = "The cluster has not drained any transactions for DRPRODUCER in last %d minutes. There are outstanding transactions."
```
Comment (xinjiacs) [inline lib/python/voltcli/checkstats.py]: @zeemanhe Please also update the message with seconds. And let Andrew know to document this new timeout shutdown stuff.

File: lib/python/voltcli/checkstats.py [hunk 6]
```diff
@@ -269,7 +335,41 @@ def check_command_log(runner):
...
+    #stats progressed, update lastUpdatedTime
+    if statsProgressed:
+        return currentTime
+
+    #stats has not made any progress since last check
+    timeSinceLastUpdate = currentTime - lastUpdatedTime
+
+    #stats timeout
+    if timeSinceLastUpdate > timeout:
+         msg = "The cluster has not drained any transactions for %s in last %d minutes. There are outstanding transactions."
```
Comment (xinjiacs) [inline lib/python/voltcli/checkstats.py]: Also here.

**Pull request merged**